From f3b27381dbbddd33fbd441dc223e8faeccf7c085 Mon Sep 17 00:00:00 2001
From: Abhishek <apatgaonkar@habana.ai>
Date: Mon, 31 Aug 2020 05:57:01 +0300
Subject: [PATCH] [SW-11870] goya2: preboot: enable uart on pldm

To get uart prints on pldm, PCIe PLL Div3 output
needs to be bumped up to 250MHz which sets actual
frequency to 12.9 KHz. Baud rate of 300 is used.

Change-Id: I07f53b2893eaa52e7b2aa6da6816f5c53e5ac08b
---

diff --git a/goya2_arc_gnu/pldm.c b/goya2_arc_gnu/pldm.c
index b2f0ff7..1acf7a6 100644
--- a/goya2_arc_gnu/pldm.c
+++ b/goya2_arc_gnu/pldm.c
@@ -10,10 +10,38 @@
 #include <preboot.h>
 #include <pll.h>
 #include <uart.h>
+#include <psoc_pci_pll_ctrl_regs.h>
+#include <psoc_global_conf_regs.h>
+#include "common.h"
 
 #define SYS_CLOCK_FREQUENCY	155000ULL
-#define UART_CLOCK_FREQUENCY	(SYS_CLOCK_FREQUENCY / 5)
-#define PLDM_UART_BAUDRATE	1200
+
+/*
+ * With Goya2 v88 PLDM image when PCI pll div3
+ * is bumped up tp 250 MHz actual clock goes
+ * to 12.9 KHz.
+ */
+#define UART_CLOCK_FREQUENCY	12900ULL
+
+/* Working BAUD rate on PLDM */
+#define PLDM_UART_BAUDRATE	300
+
+/*
+ * This cfg specific to PLDM bumps up uart freq
+ * to 250 MHz while keeping other div's on refclk
+ */
+static struct hl_pll_cfg goya2_pll_cfg[] = {
+	{
+		.name = "PCI", .base = mmPSOC_PCI_PLL_CTRL_BASE,
+		.nr = 1, .nf = 60, .od = 2, .nb = 3,
+		.div_factor = { 0, 0, 4, 1 },
+		.div_sel = { HL_PLL_REF_CLK,
+			HL_PLL_REF_CLK,
+			HL_PLL_REF_CLK,
+			HL_PLL_DIV_PLL
+		},
+	}
+};
 
 u32 hl_skip_pcie_init(void)
 {
@@ -27,6 +55,10 @@
 
 void hl_uart_init(void)
 {
+	/* Bump up UART frequency */
+	hl_pll_init(goya2_pll_cfg, 1);
+	regmap_writel(1, mmPSOC_GLOBAL_CONF_PAD_SEL_8);	/* UART0 RXD */
+	regmap_writel(1, mmPSOC_GLOBAL_CONF_PAD_SEL_9);	/* UART0 TXD */
 	preboot_uart_init(UART_CLOCK_FREQUENCY / PLDM_UART_BAUDRATE);
 }
 
diff --git a/goya2_arc_gnu/pll.c b/goya2_arc_gnu/pll.c
index ea7f16a..f19beab 100644
--- a/goya2_arc_gnu/pll.c
+++ b/goya2_arc_gnu/pll.c
@@ -7,6 +7,7 @@
 #include <stddef.h>
 #include <preboot.h>
 #include <pll.h>
+#include <uart.h>
 #include "common.h"
 
 
@@ -48,3 +49,145 @@
 	return out_freq;
 }
 
+#ifdef TARGET_IS_PLDM
+/* PLL initializations for PLDM */
+#define HL_PLL_LOCK_TOUT	10	/* 10 us */
+#define GOYA2_PLL_CFG		0x100
+#define GOYA2_PLL_CFG_EN	BIT(0)
+#define GOYA2_PLL_CFG_LOCK	BIT(16)
+#define GOYA2_UNIT_RST_N_L	0x4B0
+#define HL_PLL_DIV_FACTOR_CMD	0x220
+
+static void hl_pll_reset(struct hl_pll_cfg *cfg, bool reset)
+{
+	u32 reg = regmap_readl(cfg->base + GOYA2_PLL_CFG);
+
+	if (true == reset)
+		reg |= GOYA2_PLL_CFG_EN;
+	else
+		reg &= ~GOYA2_PLL_CFG_EN;
+
+	regmap_writel(reg, cfg->base + GOYA2_PLL_CFG);
+}
+
+static int hl_pll_is_locked(struct hl_pll_cfg *cfg)
+{
+	return regmap_readl(cfg->base + GOYA2_PLL_CFG) & GOYA2_PLL_CFG_LOCK;
+}
+
+#define GOYA2_PLL_POSTDIV1_SHIFT	16
+#define GOYA2_PLL_POSTDIV2_SHIFT	20
+#define GOYA2_PLL_REFDIV_SHIFT		24
+#define GOYA2_PLL_MAX_WAIT_TIME		1
+
+static void hl_pll_cfg(struct hl_pll_cfg *cfg)
+{
+	u32 reg = 0;
+
+	/* put the PLL into reset by writing 0 to CTRL_CFG.EN */
+	hl_pll_reset(cfg, false);
+
+	/* FBDIV - bit 0
+	 * POSTDIV1 - bit 16
+	 * POSTDIV2 - bit 20
+	 * REFDIV - bit 24
+	 */
+	reg = cfg->nr << GOYA2_PLL_REFDIV_SHIFT |
+		cfg->nb << GOYA2_PLL_POSTDIV2_SHIFT |
+		cfg->od << GOYA2_PLL_POSTDIV1_SHIFT | cfg->nf;
+
+	regmap_writel(reg, cfg->base + GOYA2_PLL_CFG_DIV);
+
+	/* after updating the coefficients we need to wait >=1us */
+	udelay(GOYA2_PLL_MAX_WAIT_TIME);
+
+	/* enable the PLL by writing a value of 1 to CTRL_CFG.EN*/
+	hl_pll_reset(cfg, true);
+}
+
+static void hl_pll_clk_switch_to_bypass(struct hl_pll_cfg *cfg)
+{
+	int sel;
+	u32 reg;
+
+	for (sel = 0; sel < HL_PLL_DIV_SIZE; sel++) {
+		reg = cfg->base + HL_PLL_DIV_SEL + (sel * sizeof(u32));
+		regmap_writel(HL_PLL_REF_CLK, reg);
+	}
+}
+
+static void hl_pll_clk_switch_to_pll(struct hl_pll_cfg *cfg)
+{
+	int sel;
+	u32 reg;
+
+	for (sel = 0; sel < HL_PLL_DIV_SIZE; sel++) {
+		reg = cfg->base + HL_PLL_DIV_SEL + (sel * sizeof(u32));
+		if (cfg->div_sel[sel] < 0 || cfg->div_sel[sel] > HL_PLL_DIV_PLL)
+			continue;
+		regmap_writel(cfg->div_sel[sel], reg);
+	}
+}
+
+static void hl_pll_wait_for_lock(struct hl_pll_cfg *cfg)
+{
+	int tout = HL_PLL_LOCK_TOUT;
+
+	while (!hl_pll_is_locked(cfg) && tout--)
+		udelay(1);
+
+	if (tout < 0) {
+		preboot_puts_raw("PLL lock timedout for");
+		preboot_puts(cfg->name);
+	}
+}
+
+static void hl_pll_div_init(struct hl_pll_cfg *cfg)
+{
+	int sel;
+	u32 div_base;
+
+	for (sel = 0; sel < HL_PLL_DIV_SIZE; sel++) {
+		if (cfg->div_sel[sel] == HL_PLL_REF_CLK ||
+		    cfg->div_factor[sel] < 0)
+			continue;
+
+		div_base = cfg->base + (sel * sizeof(u32));
+
+		regmap_writel(cfg->div_factor[sel],
+				div_base + HL_PLL_DIV_FACTOR);
+		regmap_writel(1, div_base + HL_PLL_DIV_FACTOR_CMD);
+		regmap_writel(1, div_base + HL_PLL_DIV_EN);
+	}
+
+	/* just to be on a safe side, wait for DIV init to propagate */
+	udelay(1);
+}
+
+static void hl_pll_set(struct hl_pll_cfg *cfg)
+{
+	hl_pll_clk_switch_to_bypass(cfg);
+
+	hl_pll_cfg(cfg);
+
+	/* Polling on PLL lock until lock is cleared or timed out */
+	hl_pll_wait_for_lock(cfg);
+
+	hl_pll_div_init(cfg);
+
+	hl_pll_clk_switch_to_pll(cfg);
+
+}
+
+/**
+ * hl_pll_init - Initialize pll registers
+ */
+void hl_pll_init(struct hl_pll_cfg *hl_pll_cfg, int size)
+{
+	int i;
+
+	for (i = 0; i < size; i++)
+		hl_pll_set(hl_pll_cfg + i);
+}
+
+#endif /* TARGET_IS_PLDM */
